contrib/adaptive-compression/adapt.c

```diff
@@ -0,0 +1,1011 @@
+        }
+        pthread_mutex_lock(&ctx->jobCompressed_mutex.pMutex);
+        ctx->jobCompressedID++;
+        pthread_cond_broadcast(&ctx->jobCompressed_cond.pCond);
+        pthread_mutex_unlock(&ctx->jobCompressed_mutex.pMutex);
+        if (job->lastJobPlusOne == currJob + 1 || ctx->threadError) {
+            /* finished compressing all jobs */
+            break;
+        }
+        DEBUG(2, "finished compressing job %u\n", currJob);
+        currJob++;
+    }
+    return arg;
+}
+
+static void displayProgress(unsigned cLevel, unsigned last)
+{
+    if (!g_useProgressBar) return;
+    UTIL_time_t currTime;
+    UTIL_getTime(&currTime);
+    double const timeElapsed = (double)(UTIL_getSpanTimeMicro(g_ticksPerSecond, g_startTime, currTime) / 1000.0);
+    double const sizeMB = (double)g_streamedSize / (1 << 20);
+    double const avgCompRate = sizeMB * 1000 / timeElapsed;
+    fprintf(stderr, "\r| Comp. Level: %2u | Time Elapsed: %7.2f s | Data Size: %7.1f MB | Avg Comp. Rate: %6.2f MB/s |", cLevel, timeElapsed/1000.0, sizeMB, avgCompRate);
+    if (last) {
+        fprintf(stderr, "\n");
+    }
+    else {
+        fflush(stderr);
+    }
+}
+
+static void* outputThread(void* arg)
+{
+    outputThreadArg* const otArg = (outputThreadArg*)arg;
+    adaptCCtx* const ctx = otArg->ctx;
+    FILE* const dstFile = otArg->dstFile;
+
+    unsigned currJob = 0;
+    for ( ; ; ) {
+        unsigned const currJobIndex = currJob % ctx->numJobs;
+        jobDescription* const job = &ctx->jobs[currJobIndex];
+        DEBUG(2, "starting write for job %u\n", currJob);
+        pthread_mutex_lock(&ctx->jobCompressed_mutex.pMutex);
+        while (currJob + 1 > ctx->jobCompressedID && !ctx->threadError) {
+            pthread_mutex_lock(&ctx->compressionCompletion_mutex.pMutex);
+            /* write thread is waiting on compression thread */
+            ctx->writeWaitCompressionCompletion = ctx->compressionCompletion;
+            DEBUG(2, "writer thread waiting for nextJob: %u, writeWaitCompressionCompletion %f\n", currJob, ctx->writeWaitCompressionCompletion);
+            pthread_mutex_unlock(&ctx->compressionCompletion_mutex.pMutex);
+            pthread_cond_wait(&ctx->jobCompressed_cond.pCond, &ctx->jobCompressed_mutex.pMutex);
+        }
+        pthread_mutex_unlock(&ctx->jobCompressed_mutex.pMutex);
+
+        /* reset write completion */
+        pthread_mutex_lock(&ctx->writeCompletion_mutex.pMutex);
+        ctx->writeCompletion = 0;
+        pthread_mutex_unlock(&ctx->writeCompletion_mutex.pMutex);
+
+        {
+            size_t const compressedSize = job->compressedSize;
+            size_t remaining = compressedSize;
+            if (ZSTD_isError(compressedSize)) {
+                DISPLAY("Error: an error occurred during compression\n");
+                signalErrorToThreads(ctx);
+                return arg;
+            }
+            {
+                size_t const blockSize = MAX(compressedSize >> 7, 1 << 10);
```

Comment:
for this write unit size, as well as `readBlockSize` I wasn't sure how granular I would need to make the mini-chunks in order to get useful readings for the completion %s. I eventually decided to make them approximately equal to 1/128th of the job size (or compressed size for write) since then each mini-chunk would be approximately 1% of the total job. `1 << 10` is simply a kilobyte, which I arbitrarily chose to avoid the case where `compressedSize >> 10 = 0`